### PR TITLE
Phala stakepool v2

### PR DESCRIPTION
**Pull request by Index0011**

### CODE REVIEW

File: pallets/phala/src/stakepool.rs [hunk 1]
```diff
@@ -1110,11 +1032,181 @@ pub mod pallet {
...
+
+		#[frame_support::transactional]
+		fn merge_or_init_nft_for_staker(
+			staker: T::AccountId,
+			pid: u64,
+		) -> Result<NftId, DispatchError> {
+			let pool_info = Self::ensure_pool(pid)?;
+			let collection_id =
+				PoolCollections::<T>::get(pid).ok_or(Error::<T>::MissCollectionId)?;
+			let nftid_arr: Vec<NftId> =
```
Comment (h4x3rotab) [inline pallets/phala/src/stakepool.rs:1154]: Iterate with this: https://github.com/paritytech/substrate/blob/0fbf917676824cda9c3cf7e71e83b20c26b5f92e/frame/uniques/src/lib.rs#L173-L182

File: pallets/phala/src/stakepool.rs [hunk 2]
```diff
@@ -1110,11 +1032,181 @@ pub mod pallet {
...
+			nft_id: NftId,
+			nft: &mut NftAttr<BalanceOf<T>>,
+		) -> Option<(BalanceOf<T>, BalanceOf<T>, BalanceOf<T>)> {
+			let price = pool_info.share_price()?;
+			let amount = bmul(shares, &price);
+
+			let collection_id = PoolCollections::<T>::get(pool_info.pid.clone())?;
+			let nft_id =
+				Self::merge_or_init_nft_for_staker(userid.clone(), pool_info.pid.clone()).ok()?;
+			let mut nft = Self::get_NftAttr(collection_id, nft_id).ok()?;
```
Comment (h4x3rotab) [inline pallets/phala/src/stakepool.rs:1074]: Already has `nft`

File: pallets/phala/src/stakepool.rs [hunk 3]
```diff
@@ -1186,10 +1284,11 @@ pub mod pallet {
 		/// 2. The pool must has shares and stake (or it can cause division by zero error)
 		fn try_withdraw(
 			pool_info: &mut PoolInfo<T::AccountId, BalanceOf<T>>,
-			user_info: &mut UserStakeInfo<T::AccountId, BalanceOf<T>>,
+			nft: &mut NftAttr<BalanceOf<T>>,
+			nft_id: NftId,
+			userid: T::AccountId,
```
Comment (h4x3rotab) [inline pallets/phala/src/stakepool.rs:1289]: NFT-based withdraw

File: pallets/phala/src/stakepool.rs [hunk 4]
```diff
@@ -1207,17 +1306,24 @@ pub mod pallet {
 			let (queued_shares, _) = extract_dust(queued_shares);
 			// Try withdraw immediately if we can
 			if withdrawing_shares > Zero::zero() {
-				Self::maybe_settle_slash(pool_info, user_info);
+				Self::maybe_settle_nft_slash(pool_info, nft, userid.clone());
```
Comment (h4x3rotab) [inline pallets/phala/src/stakepool.rs:1309]: A unified function to handle actual withdraw (prechecked)

File: pallets/phala/src/stakepool.rs [hunk 5]
```diff
@@ -866,26 +820,16 @@ pub mod pallet {
...
-			pool_info.add_stake(&mut user_info, a);
+			let collection_id = PoolCollections::<T>::get(pool_info.pid.clone())
+				.ok_or(Error::<T>::MissCollectionId)?;
+			let nft_id = Self::merge_or_init_nft_for_staker(who.clone(), pool_info.pid.clone())?;
+			let mut nft = Self::get_NftAttr(collection_id, nft_id)?;
+
+			Self::maybe_settle_nft_slash(&pool_info, &mut nft, who.clone());
+
+			let shares =
+				Self::add_stake_and_nft(&mut pool_info, who.clone(), collection_id, amount);
```
Comment (h4x3rotab) [inline pallets/phala/src/stakepool.rs:831]: Update nft attribute

File: pallets/phala/src/stakepool.rs [hunk 6]
```diff
@@ -429,16 +449,32 @@ pub mod pallet {
...
 			PoolCount::<T>::put(pid + 1);
+
+			let symbol: BoundedVec<u8, <T as pallet_rmrk_core::Config>::CollectionSymbolLimit> =
+				(String::from("stakepool-collection-symbol"))
+					.as_bytes()
+					.to_vec()
+					.try_into()
+					.unwrap();
+			let meta: BoundedVec<u8, <T as pallet_uniques::Config>::StringLimit> =
+				(String::from("stakepool-collection-phala:") + pid.to_string().as_str())
```
[the remark this answers is not in the document]
Comment (Index0011) [reply pallets/phala/src/stakepool.rs:469]: done

File: pallets/phala/src/stakepool.rs [hunk 7]
```diff
@@ -429,16 +449,32 @@ pub mod pallet {
...
 				},
 			);
 			PoolCount::<T>::put(pid + 1);
+
+			let symbol: BoundedVec<u8, <T as pallet_rmrk_core::Config>::CollectionSymbolLimit> =
+				(String::from("stakepool-collection-symbol"))
+					.as_bytes()
+					.to_vec()
+					.try_into()
+					.unwrap();
```
[the remark this answers is not in the document]
Comment (Index0011) [reply pallets/phala/src/stakepool.rs:467]: done

File: pallets/phala/src/stakepool.rs [hunk 8]
```diff
@@ -429,16 +449,32 @@ pub mod pallet {
...
+			let meta: BoundedVec<u8, <T as pallet_uniques::Config>::StringLimit> =
+				(String::from("stakepool-collection-phala:") + pid.to_string().as_str())
+					.as_bytes()
+					.to_vec()
+					.try_into()
+					.unwrap();
+			pallet_rmrk_core::Pallet::<T>::create_collection(origin, meta, None, symbol)?;
```
[the remark this answers is not in the document]
Comment (Index0011) [reply pallets/phala/src/stakepool.rs:474]: done

File: pallets/phala/src/stakepool.rs (hunk 1 above)
Comment (Index0011) [reply pallets/phala/src/stakepool.rs:1154]: waiting for joshua's reply

File: pallets/phala/src/stakepool.rs (hunk 3 above)
Comment (Index0011) [reply pallets/phala/src/stakepool.rs:1289]: done

File: pallets/phala/src/stakepool.rs (hunk 2 above)
Comment (Index0011) [reply pallets/phala/src/stakepool.rs:1074]: done

File: pallets/phala/src/stakepool.rs (hunk 5 above)
Comment (Index0011) [reply pallets/phala/src/stakepool.rs:831]: done

File: pallets/phala/src/stakepool.rs [hunk 9]
```diff
@@ -1110,11 +1032,181 @@ pub mod pallet {
...
+			let collection_id =
+				PoolCollections::<T>::get(pid).ok_or(Error::<T>::MissCollectionId)?;
+			pallet_rmrk_core::Pallet::<T>::set_lock((collection_id, nft_id), false);
+			pallet_rmrk_core::Pallet::<T>::burn_nft(
+				Origin::<T>::Signed(pool_info.owner.clone()).into(),
+				collection_id,
+				nft_id,
+			)?;
+
+			Ok(())
```
[the remark this answers is not in the document]
Comment (Index0011) [reply pallets/phala/src/stakepool.rs:1143]: wait for joshua's reply

File: pallets/phala/src/migrations/mod.rs [hunk 10]
```diff
@@ -5,6 +5,10 @@ use frame_support::{
 };
 use log;
 
+pub use pallet_rmrk_core::types::*;
+pub use pallet_rmrk_market;
+pub use rmrk_traits::primitives::*;
```
Comment (h4x3rotab) [inline pallets/phala/src/migrations/mod.rs:10]: Remove unused use

**Comment by h4x3rotab** [comment]
Will move to the new PR to continue the review

**Pull request merged**